FormalConjectures/Wikipedia/EllipticCurveRank.lean

```diff
@@ -0,0 +1,205 @@
+
+open WeierstrassCurve in
+instance (E : RatEllipticCurve) : E.toWeierstrass.IsElliptic where
+  isUnit := isUnit_iff_ne_zero.mpr <| by
+    convert mul_ne_zero (show (-16 : ℚ) ≠ 0 by norm_num) (Int.cast_ne_zero.mpr E.Δ_ne_zero)
+    simp_rw [toWeierstrass, Δ, b₂, b₄, b₆, Int.cast_add, Int.cast_mul, Int.cast_pow]
+    ring
+
+/-- The naïve height of an elliptic curve over ℚ. -/
+def naiveHeight (E : RatEllipticCurve) : ℕ := max (4 * E.A.natAbs ^ 3) (27 * E.B.natAbs ^ 2)
+
+/-- The set of elliptic curves over ℚ with naïve height less than or equal to a given height. -/
+def heightLE (H : ℕ) : Set RatEllipticCurve := {E : RatEllipticCurve | E.naiveHeight ≤ H}
+
+open scoped Topology
+open Filter (atTop)
+
+/-- Conjecture by Goldfeld and Katz–Sarnak: if elliptic curves over ℚ are ordered by their
+heights, then 50% of the curves have rank 0 and 50% have rank 1.
+See p. 28 of https://people.maths.bris.ac.uk/~matyd/BSD2011/bsd2011-Bhargava.pdf. -/
+@[category research open, AMS 11 14]
+theorem half_rank_zero_and_half_rank_one (r : ℕ) (hr : r = 0 ∨ r = 1) :
+    atTop.Tendsto
+      (fun H ↦ ({E ∈ heightLE H | E.rank = r}.ncard / (heightLE H).ncard : ℝ)) (𝓝 (1 / 2)) := by
+  sorry
+
+/-- Theorem 3 of [BS2013]:
+when elliptic curves over ℚ are ordered by height, their average rank is < .885. -/
+@[category research solved, AMS 11 14]
+theorem avg_rank_lt_0885 :
+    atTop.limsup (fun H ↦ ((∑ᶠ E : heightLE H, E.1.rank) / (heightLE H).ncard : ℝ)) < 0.885 := by
+  sorry
+
+/-- Theorem 4 of [BS2013]:
+when elliptic curves over ℚ are ordered by height, a density of at least 83.75% have
+rank 0 or 1. -/
+@[category research solved, AMS 11 14]
+theorem _08375_le_density_rank_zero_one : 0.8375 ≤ atTop.liminf
+    fun H ↦ ({E ∈ heightLE H | E.rank = 0 ∨ E.rank = 1}.ncard / (heightLE H).ncard : ℝ) := by
+  sorry
+
+/-- Theorem 5 of [BS2013]:
+when elliptic curves over ℚ are ordered by height, a density of at least 20.62% have rank 0. -/
+@[category research solved, AMS 11 14]
+theorem _02062_le_density_rank_zero : 0.2062 ≤ atTop.liminf
+    fun H ↦ ({E ∈ heightLE H | E.rank = 0}.ncard / (heightLE H).ncard : ℝ) := by
+  sorry
+
+/-- From [PPVW2016], Section 3.1: "from the mid-1960s to the present,
+it seems that most experts conjectured unboundedness". -/
+@[category research open, AMS 11 14]
+theorem unbounded_rank_conjecture (n : ℕ) : ∃ E : RatEllipticCurve, n ≤ E.rank := by
+  sorry
+
+/-- From [PPVW2016], Section 8.2:
+"Our heuristic predicts (a) All but finitely many E ∈ ℰ satisfy rk E(ℚ) ≤ 21".
+In other words, there are only finitely many elliptic curves over ℚ (up to isomorphism)
+with rank greater than 21.
+Notice that this contradicts the previous conjecture. -/
+@[category research open, AMS 11 14]
+theorem finite_twentyone_lt_finrank : {E : RatEllipticCurve | 21 < E.rank}.Finite := by
+  sorry
+
+/-- [PPVW2016] 8.2(b). Note: ℰ_H should be ℰ_{≤H}, see Theorem 7.3.3. -/
```

Comment:
For this one and the one below, I think it would also be worth having the informal statements in the docstring.